hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/JpaDistributionSetType.java

```diff
@@ -209,8 +210,12 @@ public void setKey(final String key) {
 
     @Override
     public boolean checkComplete(final DistributionSet distributionSet) {
-        return distributionSet.getModules().stream().map(SoftwareModule::getType).collect(Collectors.toList())
-                .containsAll(getMandatoryModuleTypes());
+        List<SoftwareModuleType> smTypes = distributionSet.getModules().stream().map(SoftwareModule::getType)
```

Comment:
could be simplified:
```
final Set<SoftwareModuleType> mandatoryModuleTypes = getMandatoryModuleTypes();
return distributionSet.getModules().stream().map(SoftwareModule::getType).allMatch(type -> mandatoryModuleTypes.contains(type));
```

Reply:
This will not work in 1 specific case. When distributionSet.getModules() is empty and mandatoryModuleTypes is also empty then allMatch returns true. I just tried it.

I guess that allMatch assumes true and looks for a false. And in above, no iteration happens.

Reply:
You are right, we can avoid that by adding:
```
        if (smTypes.isEmpty()) {
            return false;
        }
```
before the stream call

Reply:
I asume you are referring to something like this:
```
 if (distributionSet.getModules().isEmpty()) {
            return false;
        }
        final Set<SoftwareModuleType> mandatoryModuleTypes = getMandatoryModuleTypes();
        return distributionSet.getModules().stream().map(SoftwareModule::getType)
                .allMatch(type -> mandatoryModuleTypes.contains(type));
```

This has a problem. It returns false even when I assign a software module to this type. In theory, this is a case where a DS type has no mandatory SM type. But it has a SM assigned. So it must be marked complete right?

Reply:
Its for 2 resons:
- An incomplete DS should not be visible in deployment view.
- If all software modules are removed from a given DS, then this DS should be marked incomplete.

Reply:
Ok, now I got it.
Maybe a good idea here is to introduce some unit tests? specially if a Model class has some logic like this

Reply:
You can maybe use the `DistributionSetTypeManagementTest` test, and test the different use  cases  where this method is called, here it's called in `constructor`, `addModule`, and `removeModule`

Reply:
Discussed with @stefbehl to add the test cases later as an improvement. The tests will be added in the next sprint.